### ISSUE TITLE
[Feature] allow viewport size to be independent from the window size

### ISSUE THREAD

**Issue by idoros** [microsoft/playwright#2749]
I'm trying to test for responsive behavior of my library, The tests inspect the HTML state in a specific size, resize the viewport and re-check the state. 

First of all, for debugging, I really like the behavior of the fixed viewport which allows debugging with a full window while keeping the document viewport restricted.

However there are several issues with the way the viewport works that makes it hard to build responsive tests:
1. once viewport is set, there is no way to return to `viewport: null` ([setViewportSize doesn't accept null](https://github.com/microsoft/playwright/blob/master/docs/api.md#pagesetviewportsizeviewportsize)).
2. changing the viewport size also triggers a change to the window size - that makes it hard to run multiple tests under the same window and removes the nice benefit that the separation of window & viewport sizes gives. It is also not consistent across browsers (Chromium changes size and if the dev tools are open it might minimize to hide the viewport while Firefox only changes the viewport and doesn't change the window), it would be nice to maybe add a flag to prevent changing the window size.
3. There is no API to control the window size (although it exist under the devTools protocol: [setWindowBounds](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-setWindowBounds), [getWindowBounds](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-getWindowBounds)).

Giving control over the window size and viewport size would really help building maintainable tests for libraries, applications and websites resize behavior.

**Comment by aslushnikov** [comment]
> Giving control over the window size and viewport size would really help building maintainable tests for libraries, applications and websites resize behavior.

@idoros how's window size used in your libraries code? For example, I was under impression that MediaQueries work against viewport size or screen size, so window size is not used.

Am I missing something?

**Comment by idoros** [comment]
Its not. Normally the viewport or documentElement is bound to the window inner edges. It's just under Playwright, or dev tools protocol, that allows to separate the viewport from the window.

From the perspective of the code being tested, there is no different between the viewport and the window.

My issue is about the inconsistency of offering the control of window size with the setViewportSize api (which separates the viewport size from the window size, I can only assume that it's for debugging purposes) and then changing the window size when the viewport size changes.

**Comment by dgozman** [comment]
We made window size mimic viewport size for the ease of debugging. Usually, it is much easier to visually see the changed window size instead of the more subtle inner viewport size. Are you asking for a way to disable this feature?

I'd be interested in your workflow. Why do you want the window size stay the same? Do you have DevTools docked in that window? Wouldn't you want to undock it? Also, as you mention, this is very browser-specific. For example, WebKit cannot resize just the viewport, so resizing the window is inevitable there.

> that makes it hard to run multiple tests under the same window

What do you think about creating a separate context for each test, so that they do not share the window?

**Comment by idoros** [comment]
I chose to run my entire suit of tests in a single context/page because it's much faster and easier to test and debug. Running in the browser, allows tests to have direct sync access to the document with each test easily setting up and tearing down its own fixture. While I provide async access to the Playwright API only to mimic user actions, collect external information or save snapshots ([click example](https://github.com/idoros/zeejs/blob/master/packages/react/test/test.spec.tsx#L282)).

Working with the browser open allows me to run code and debug it directly in the browser while still having the great external control and information that Playwright provides.

In dev mode I run my IDE on one window and the browser (viewport + dev tools) on the other. Having the window change size interrupt the dev experience.

Disabling the feature would be great! but also giving access to the dev tools protocol window bounds API could help. 

It's true that the abilities diff between browsers and I can understand that it's not simple or even possible to provide an identical DX (dev tools don't even work with webkit; not sure if it's only on my machine). But I mostly debug with chromium and I find the current behavior problematic.

**Comment by dgozman** [comment]
> I chose to run my entire suit of tests in a single context/page because it's much faster and easier to test and debug. 

Oh, I was under impression that you had multiple pages, was reading it wrong 😄  

> In dev mode I run my IDE on one window and the browser (viewport + dev tools) on the other. Having the window change size interrupt the dev experience.

This makes sense. Let me rephrase the issue to reflect the findings. Overall, your setup seems to be "run my unit tests in the browser controlled by Playwright". This is somewhat different from our end-to-end focus, so most likely there will be rough edges in this case, and somewhat slower prioritization.

> Disabling the feature would be great! but also giving access to the dev tools protocol window bounds API could help.

I don't think we'll go for the window bounds api in the nearest future - we strongly prefer browser-agnostic solutions, and this one would be hard to make consistent. That said, disabling this feature in Chromium should work for you - you'll be able to resize the window manually.

**Comment by idoros** [comment]
> disabling this feature in Chromium should work for you - you'll be able to resize the window manually.

That would be awesome 

> somewhat slower prioritization

Will it be faster if I provide the PR?

**Comment by dgozman** [comment]
> > somewhat slower prioritization
> 
> Will it be faster if I provide the PR?

PRs are always welcome! I am not yet sure what the api would be. Maybe a second parameter to `page.setViewportSize()`?

**Comment by idoros** [comment]
I thought of a second param to `page.setViewportSize()` as well, alternatively it could be a browser option which will allow to set  the behavior according to the test env.

Should the PR change the Firefox default behavior to modify the window size according to the viewport size?

**Comment by dgozman** [comment]
> Should the PR change the Firefox default behavior to modify the window size according to the viewport size, or to leave it for a separate PR?

Up to you 😄

**Comment by idoros** [comment]
I'm checking out what needs/can to be done, and I wonder if you can guide me a bit since its my first PR in Playwright.

**Testing:**
1. I assume testing for the new flag should be in `emulation.spec`
2. should I just create a new test that runs only on Chromium at first?
3. not sure if the tests are permitted to go into the page internals in order to get things like actual window size (and since it's for testing only I don't think adding `page.getIndependentWindowSize()` method would be accepted)
	- while checking for sizes withing the browser I found out that `window.outerWidth/outerHeight` gives the real window size on Chromium and the viewport size on Firefox (not actually sure what would be result I would expect, but it might be a real bug in Chromium or Firefox)
	- should I test with `window.outerWidth/outerHeight` on chromium? or can you suggest another way of testing this? or not test this at all (I can see that the previous change for the window size didn't test this)
4. cloning Playwright without changing anything gave me 10 broken tests (1 in Firefox `defaultbrowsercontext.spec`, 4 in Firefox `headful.spec`, 2 in FireFox `multiclient.spec`, 3 in WebKit `navigation.spec`), I assume It's a "my machine" issue that will work on CI

**Docs:**
1. Adding the description for this flag, should I expand on the various window behaviors between the browsers?

**RPC** - I see there is a new RPC API under src, but there are not tests, should I just add the additional arg to the relevent signatures and code?

-----

> We made window size mimic viewport size for the ease of debugging. Usually, it is much easier to visually see the changed window size instead of the more subtle inner viewport size.

btw, changing viewport size without changin window size gives the area outsize the viewport a black color on Chromium

**Comment by dgozman** [comment]
> I'm checking out what needs/can to be done, and I wonder if you can guide me a bit since its my first PR in Playwright.

Thank you for contributing!

>    * should I test with `window.outerWidth/outerHeight` on chromium? or can you suggest another way of testing this? or not test this at all (I can see that the previous change for the window size didn't test this)

I think this is the easiest way. Just check outerWidth/outerHeight before/after calling the method with and without the flag. `emulation.spec.js` is the right place. Mark it as `it.skip(!CHROMIUM)`.

> 4. cloning Playwright without changing anything gave me 10 broken tests (1 in Firefox `defaultbrowsercontext.spec`, 4 in Firefox `headful.spec`, 2 in FireFox `multiclient.spec`, 3 in WebKit `navigation.spec`), I assume It's a "my machine" issue that will work on CI

Oh, that's bad news. Mind sharing the list of failing tests? We are serious about fighting flakiness.

> **Docs:** - Adding the description for this flag, should I expand on the various window behaviors between the browsers?

Sure, why not. I think it would be easier to discuss in the PR itself.
 
> **RPC** - I see there is a new RPC API under src, but there are no tests, should I just add the additional arg to the relevant signatures and code?

Tests for rpc could be run with `PWCHANNEL=1 npm run test`. This is still experimental, but plumbing it through will be the right thing.

**Comment by idoros** [comment]
I opened a [PR](https://github.com/microsoft/playwright/pull/2815).

-----

About the failing tests:
1. I get 2 logs in the console with `ERROR: the process "XXXX" not found when running only on Chromium (not getting them when running with `PWCHANNEL=1`)
2. The failing test log when running everything normally:
```
Failures:
1) [TIMEOUT 10000ms] Firefox launchPersistentContext() should support viewport option (defaultbrowsercontext.spec.js:136:3)
 Output:
 [extra] [browser] => browserType.launchPersistentContext started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -headless -profile C:\Users\mmloc\AppData\Local\Tem
p\playwright_dev_profile-GLLstk -juggler 0 about:blank
 [extra] [browser] <launched> pid=19280
 [extra] [browser] *** You are running in headless mode.

2) [TIMEOUT 10000ms] Firefox Headful should not crash when creating second context (headful.spec.js:60:3)
 Output:
 [extra] [browser] => browserType.launch started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -wait-for-browser -foreground -profile C:\Users\mml
oc\AppData\Local\Temp\playwright_firefoxdev_profile-GalIOO -juggler 0 -silent
 [extra] [browser] <launched> pid=20792
 [extra] [browser] Juggler listening on ws://127.0.0.1:54806/e1748ff6-e646-440a-ad24-9d9b7f172414
 [extra] [browser] <ws connecting> ws://127.0.0.1:54806/e1748ff6-e646-440a-ad24-9d9b7f172414
 [extra] [browser] <ws connected> ws://127.0.0.1:54806/e1748ff6-e646-440a-ad24-9d9b7f172414
 [extra] [browser] <= browserType.launch succeeded
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
 [extra] [browser] [Parent 25248, Gecko_IOThread] WARNING: file c:/playwright/browser_patches/firefox/checkout/ipc/chromium/src/base/process_util_win.cc, line 166
 [extra] [browser] [Parent 25248, Gecko_IOThread] WARNING: file c:/playwright/browser_patches/firefox/checkout/ipc/chromium/src/base/process_util_win.cc, line 166

3) [TIMEOUT 10000ms] Firefox Headful should click background tab (headful.spec.js:74:3)
 Output:
 [extra] [browser] => browserType.launch started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -wait-for-browser -foreground -profile C:\Users\mml
oc\AppData\Local\Temp\playwright_firefoxdev_profile-LUpwgc -juggler 0 -silent
 [extra] [browser] <launched> pid=6836
 [extra] [browser] Juggler listening on ws://127.0.0.1:54828/235d0deb-5ea3-47ff-ac67-0d3d6e6401e6
 [extra] [browser] <ws connecting> ws://127.0.0.1:54828/235d0deb-5ea3-47ff-ac67-0d3d6e6401e6
 [extra] [browser] <ws connected> ws://127.0.0.1:54828/235d0deb-5ea3-47ff-ac67-0d3d6e6401e6
 [extra] [browser] <= browserType.launch succeeded
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.

4) [TIMEOUT 10000ms] Firefox Headful should(not) block third party cookies (headful.spec.js:82:3)
 Output:
 [extra] [browser] => browserType.launch started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -wait-for-browser -foreground -profile C:\Users\mml
oc\AppData\Local\Temp\playwright_firefoxdev_profile-paRr23 -juggler 0 -silent
 [extra] [browser] <launched> pid=14908
 [extra] [browser] Juggler listening on ws://127.0.0.1:54844/cd929323-e81a-46df-9332-648b304705bf
 [extra] [browser] <ws connecting> ws://127.0.0.1:54844/cd929323-e81a-46df-9332-648b304705bf
 [extra] [browser] <ws connected> ws://127.0.0.1:54844/cd929323-e81a-46df-9332-648b304705bf
 [extra] [browser] <= browserType.launch succeeded
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.

5) [TIMEOUT 10000ms] Firefox Headful should not override viewport size when passed null (headful.spec.js:121:18)
 Output:
 [extra] [browser] => browserType.launch started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -wait-for-browser -foreground -profile C:\Users\mml
oc\AppData\Local\Temp\playwright_firefoxdev_profile-6vSUO3 -juggler 0 -silent
 [extra] [browser] <launched> pid=22788
 [extra] [browser] Juggler listening on ws://127.0.0.1:54866/559d9d5d-5d47-4e11-81b0-42f08b1f4d4f
 [extra] [browser] <ws connecting> ws://127.0.0.1:54866/559d9d5d-5d47-4e11-81b0-42f08b1f4d4f
 [extra] [browser] <ws connected> ws://127.0.0.1:54866/559d9d5d-5d47-4e11-81b0-42f08b1f4d4f
 [extra] [browser] <= browserType.launch succeeded
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.

6) [TIMEOUT 10000ms] Firefox BrowserContext should work across sessions (multiclient.spec.js:21:3)
 Output:
 [extra] [browser] => browserType.launchServer started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -headless -profile C:\Users\mmloc\AppData\Local\Tem
p\playwright_firefoxdev_profile-UcjwFT -juggler 0 -silent
 [extra] [browser] <launched> pid=19204
 [extra] [browser] *** You are running in headless mode.
 [extra] [browser] Juggler listening on ws://127.0.0.1:54893/6ed2bdb0-f73b-4549-9a74-ad1aa175ad68
 [extra] [browser] <ws connecting> ws://127.0.0.1:54893/6ed2bdb0-f73b-4549-9a74-ad1aa175ad68
 [extra] [browser] <ws connected> ws://127.0.0.1:54893/6ed2bdb0-f73b-4549-9a74-ad1aa175ad68
 [extra] [browser] <= browserType.launchServer succeeded
 [extra] [browser] <gracefully close start>
 [extra] [browser] <ws disconnected> ws://127.0.0.1:54893/6ed2bdb0-f73b-4549-9a74-ad1aa175ad68
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
 [extra] [browser] <process did exit: exitCode=0, signal=null>

7) [TIMEOUT 10000ms] Firefox browserType.connect should be able to connect multiple times to the same browser (multiclient.spec.js:80:3)
 Output:
 [extra] [browser] => browserType.launchServer started
 [extra] [browser] <launching> C:\Users\mmloc\AppData\Local\ms-playwright\firefox-1116\firefox\firefox.exe -no-remote -headless -profile C:\Users\mmloc\AppData\Local\Tem
p\playwright_firefoxdev_profile-z8pDth -juggler 0 -silent
 [extra] [browser] <launched> pid=23944
 [extra] [browser] *** You are running in headless mode.
 [extra] [browser] Juggler listening on ws://127.0.0.1:54924/7ae9af5d-880b-4695-be82-80c9a9551931
 [extra] [browser] <ws connecting> ws://127.0.0.1:54924/7ae9af5d-880b-4695-be82-80c9a9551931
 [extra] [browser] <ws connected> ws://127.0.0.1:54924/7ae9af5d-880b-4695-be82-80c9a9551931
 [extra] [browser] <= browserType.launchServer succeeded
 [extra] [browser] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
 [extra] [browser] [Parent 25088, Gecko_IOThread] WARNING: file c:/playwright/browser_patches/firefox/checkout/ipc/chromium/src/base/process_util_win.cc, line 166
 [extra] [browser] <gracefully close start>
 [extra] [browser] <ws disconnected> ws://127.0.0.1:54924/7ae9af5d-880b-4695-be82-80c9a9551931

8) [FAIL] WebKit Page.goto should fail when navigating to bad SSL (navigation.spec.js:154:3)

 1035 | } else if (WEBKIT) {
 1036 | if (MAC)
 1037 | expect(errorMessage).toContain('The certificate for this server is invalid');
 1038 | else if (WIN)
 1039 | expect(errorMessage).toContain('SSL peer certificate or SSH remote key was not OK');
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^

 Received: "SSL connect error\n====================== page.goto logs ======================\n[api] navigating to \"https://localhost:8908/empty.html\", waiting until \
"load\"\n============================================================\nNote: use DEBUG=pw:api environment variable and rerun to capture Playwright logs."


9) [FAIL] WebKit Page.goto should fail when navigating to bad SSL after redirects (navigation.spec.js:164:3)

 1035 | } else if (WEBKIT) {
 1036 | if (MAC)
 1037 | expect(errorMessage).toContain('The certificate for this server is invalid');
 1038 | else if (WIN)
 1039 | expect(errorMessage).toContain('SSL peer certificate or SSH remote key was not OK');
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^

 Received: "SSL connect error\n====================== page.goto logs ======================\n[api] navigating to \"https://localhost:8908/redirect/1.html\", waiting un
til \"load\"\n============================================================\nNote: use DEBUG=pw:api environment variable and rerun to capture Playwright logs."


10) [FAIL] WebKit Page.waitForNavigation should work with clicking on links which do not commit navigation (navigation.spec.js:594:3)

 1035 | } else if (WEBKIT) {
 1036 | if (MAC)
 1037 | expect(errorMessage).toContain('The certificate for this server is invalid');
 1038 | else if (WIN)
 1039 | expect(errorMessage).toContain('SSL peer certificate or SSH remote key was not OK');
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^

 Received: "SSL connect error\n=============== page.waitForNavigation logs ===============\n[api] waiting for navigation until \"load\"\n==============================
==============================\nNote: use DEBUG=pw:api environment variable and rerun to capture Playwright logs."
```
